### ISSUE TITLE
Allow batched requests to subgraphs

### ISSUE THREAD

**Issue by samuelAndalon** [apollographql/router#2002]
**Is your feature request related to a problem? Please describe.**
For graph consumers is a very common scenario to have data fetching logic with the component that use it, causing many requests to a supergraph, to fix this we have the option of using batched requests to supergraphs, then the supergraph will concurrently and independently execute all operations in the batch. 

The problem with this is that, given that operations are executed independently, query planner will create a plan for each operation potentially causing a lot of requests to same subgraph

Example
<img width="607" alt="image" src="https://user-images.githubusercontent.com/6611331/197892554-18982bce-bf81-4ced-91db-897424b42702.png">

This batched request to the supergraph runtime with 2 operations will cause 4 separated network requests to subgraphs.

At Expedia Group we found a solution for this problem in the gateway-js by adding the dataloader pattern between the communication of supergraph and subgraphs by pretty much having a queue on the request context and take advantage of the javascript microtask queue to dispatch data loader

<img width="649" alt="image" src="https://user-images.githubusercontent.com/6611331/197892884-87a5cd01-4120-416a-b5e7-f805457c2665.png">

Now for the router things are more interesting given the fact that there is no event loop so a custom dispatch mechanism will need to be implemented.

**Describe the solution you'd like**
We believe that this feature is a must on supergraph given the way some clients write and execute requests against the supergraph -- performance, network optimization but also consistency are some of the reasons why we added this feature in our gateway implementation.

Now, in order to write a custom dispatch mechanism, i was thinking that maybe we can inspect each Query Plan for each operation in a batch, and have some some of expected requests vs current fired requests 

we will have n Query Plans and each Query Plan AST will have m levels
so it would be matter of check for each level how many requests we need to fire vs how many requests we have fired, i guess rust has a concept of multithreading, but applying a semaphore lock over a variable that holds the count of fired requests would help.

for graphql-kotlin we had to do something very similar to dispatch data loaders by tracking the synchronous execution of a graphQL operation, once the synchronous execution is exhausted, then we need to dispatch. 

Additional information

https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/data-loader-instrumentation#dispatching-by-synchronous-execution-exhaustion

**Comment by Geal** [comment]
the idea is interesting, but we have to be careful here, as anything that introduces synchronization has the potential to break performance catastrophically. In a proxy like the router, the happy path is when all requests can execute independently.

So, to explore the problem, I need a bit more info:
- what is the issue with the number of subgraph requests? Connection handshake latency, latency for each request, bandwidth, or having the subgraph answer multiple queries instead of one?
- is the connection to subgraphs in HTTP 1 or HTTP/2? (HTTP/2 has parallel streams over a single connection)
- would using `@defer` help? Currently the query planner does not handle batched queries, but you can achieve a similar behaviour with `@defer`, with added benefits:
  - the client receives each part of the query as soon as it is available, instead of waiting for all queries to finish
  - the query planner is able to factor subgraph requests from multiple deferred fragments, thus reducing the number of requests

**Comment by samuelAndalon** [comment]
> the happy path is when all requests can execute independently.
with the custom dispatching mechanism we did (https://github.com/samuelAndalon/space-camp-federation-demo/tree/feat/gateway-batching/src/gateway/remote-graphql-batch-datasource), technically, operations in a batch request are still being executed independently, what we added is a layer of indirection colocated in the incoming request context, to hold operations, and dispatch them using microtasks.

so the `ExecutionService` could potentially be agnostic of it, and we could colocate the dataloader pattern in between `ExecutionService` and `SubgraphService`
<img width="448" alt="image" src="https://user-images.githubusercontent.com/6611331/198183408-e8f0a876-2fdb-40a2-9c67-97a859fc9e37.png">

about your questions:
> what is the issue with the number of subgraph requests? Connection handshake latency, latency for each request, bandwidth, or having the subgraph answer multiple queries instead of one?

1. Waste of network resources, before turning on the custom mechanism we wrote on our gateway we had 2.6k tps, after turning on, requests to subgraphs decreased dramatically to 1.44 tps
![image](https://user-images.githubusercontent.com/6611331/198183857-c5f53640-f7ea-414d-bbf2-32fbc5ad9726.png)

But its not only about network resources, given that gateway is sending separate requests to subgraphs, the usage of DataLoaders and Deduplication at subgraph layer is not possible given that subgraphs are getting separate requests.

So this brings the other use case which is consistency, downstream apis, wrapped in resolvers/data fetchers might be able to resolve graphql queries/fields in 1 single transaction, even if operations are different.

for example

![image](https://user-images.githubusercontent.com/6611331/198184246-0e3dacb0-212c-42ea-ba71-52da096735a9.png)

> is the connection to subgraphs in HTTP 1 or HTTP/2? (HTTP/2 has parallel streams over a single connection)
this fully depends on how the communication is stablished between supergraph and subgraphs, and given that batched requests are optional at subgraph server layer, we also added the option to enable batching to subgraphs by subgraph.

> would using @defer help? Currently the query planner does not handle batched queries, but you can achieve a similar behavior with @defer
it could definitely help, but not for all the use cases described above, specially the consistency, we might want to resolve the content of 2 components, each one unaware of the other, in 1 single request to downstream apis.

**Comment by ahrushetskyi** [comment]
Batch queries to subgraph can also be used as a workarond for https://github.com/apollographql/router/issues/1929

**Comment by Object905** [comment]
Having pretty big performance problems due to this.
Number of subgraph requests may induce heavy performance hit for different sub-graphs and configurations that may be written in different languages or live on different networks. For example  
Also having shared data-loaders between resolvers of related entities is really helping, especially in case of 3-4 level nesting.
Dealing with authentication\authorization may be slower, because each separate query may have to fetch user\permissions again.

What would be great is config option in traffic_shaping to do something like this when it resolves entities, so all queries to 1 subparagraph are batched into single one. This seems enough for most use cases, and avoids queues/synchronization.

```graphql
query getTariffs__backend__1($representations_1: [_Any!]!, $representations_2: [_Any!]!) {
  _result_1: _entities(representations: $representations_1) {
    ... on Subscriptions {
      __typename
      name
      icon
      slug
    }
  }
  _result_2: _entities(representations: $representations_2) {
    ... on SubscriptionService {
      name
      icon
      slug
    }
  }
}
```

Not familiar with codebase, but from perspective this seems not that hard.

**Comment by Geal** [comment]
> Not familiar with codebase, but from perspective this seems not that hard

This is actually a hard problem, but not due to the codebase:
- it needs a new synchronization point on subgraph requests, which can critically reduce overall performance or introduce complex bugs. This is a topic that we consider with care
- batching can actually add a lot of latency, because the entire subgraph request has to wait for all the batched queries to finish, so with only one slow subgraph query the whole batch can suddenly be slower. So now we need a mechanism to put different queries in different buckets, depending on the use of some fields or variables
- artificially adding latency to a lot of requests, which can quickly add up when one query generates a lot of subgraph requests. If the span during which we accumulate queries to batch is too small, the batches are small and there's not much benefit, while if it's too large, a lot of queries will wait on it. So that must be configurable, and likely a different value per bucket, because we would want similar queries that are executed frequently to batch together, while less frequent ones could go alone
- this has weird interactions with caching (that we will soon look into): how would a `Cache-Control` directive apply? On the entire batch of queries? Would we keep the etag of the batch for each individual query? Then what happens when sending a batch with queries that happened in separate batches?
- this has weird interactions with authorization. What happens when the queries come with an `Authorization` header? Should we send a batch with multiple of those headers, or have a separate bucket per authorization context? What should the subgraph return? a 401 that applies to the entire batch?

Implementing subgraph request batching requires a satisfactory solution for those issues, which takes some time, and right now this is not on the roadmap.

Reducing the load on subgraphs is an important topic for us though, so there might be some other things we can do. I alread mentioned that we are looking into caching. Have you tested the deduplication optimization features? There are two of them you can try:
- variable deduplication eliminates redundant variables in federated queries, which can happen when some entities are requested from multiple parent entities. This can greatly reduce the query size and the execution time in the subgraph
- subgraph query deduplication: if the same subgraph request is already executing, we will wait for its result and copy it instead of sending another request. It matches on the entire subgraph request (path, headers and body)

You can activate them as follows:

```yaml
traffic_shaping:
  deduplicate_variables: true # Enable the variable deduplication optimization.
  all:
    deduplicate_query: true # Enable query deduplication for all subgraphs.
```

**Comment by dbanty** [comment]
I just watched @samuelAndalon's [Summit talk (short version)](https://www.youtube.com/watch?v=cIDsg0cyQEE) and came looking for an equivalent router feature. It seems to me that the requested version of this feature does not require batching multiple HTTP requests—but rather multiple operations within one request. Please correct me if I'm wrong—but I _think_ this means a lot of the challenges that @Geal pointed out go away.

> it needs a new synchronization point on subgraph requests, which can critically reduce overall performance or introduce complex bugs. This is a topic that we consider with care

This might still be true—but the sync point is at the request handler so it might just be moving some things around rather than creating _new_ sync points.

> batching can actually add a lot of latency, because the entire subgraph request has to wait for all the batched queries to finish, so with only one slow subgraph query the whole batch can suddenly be slower. So now we need a mechanism to put different queries in different buckets, depending on the use of some fields or variables

Either way this is one HTTP request, so the worst case should be the same as the current worst case—then folks should use `@defer` to optimize slow bits from there 😁 

> - artificially adding latency to a lot of requests, which can quickly add up when one query generates a lot of subgraph requests. If the span during which we accumulate queries to batch is too small, the batches are small and there's not much benefit, while if it's too large, a lot of queries will wait on it. So that must be configurable, and likely a different value per bucket, because we would want similar queries that are executed frequently to batch together, while less frequent ones could go alone
> - this has weird interactions with caching (that we will soon look into): how would a Cache-Control directive apply? On the entire batch of queries? Would we keep the etag of the batch for each individual query? Then what happens when sending a batch with queries that happened in separate batches?
> - this has weird interactions with authorization. What happens when the queries come with an Authorization header? Should we send a batch with multiple of those headers, or have a separate bucket per authorization context? What should the subgraph return? a 401 that applies to the entire batch?

None of these are issues if this is one request.

---

There are some other points in this thread that _do_ rely on batching being cross-request or request-independent, but I think that harder problem could be delayed into a separate issue to speed up the near-term gains of single-request, multiple-operation queries.
